Eager style export V0 API.

zhxchen17 opened the pull request:
Summary:
Prototype of an end-to-end export workflow to call a torch.compiled model eagerly and package every single compiled model in the wrapped region of the code.

Code sample:

```
@torch.compile(fullgraph=True)
def f(x, y):
    return x + y

# Compile the model and save it on disk
with torch.compiler._fullgraph_package(mode="package", path="/tmp/model.pt2", frontend="dynamo"):
    f(torch.randn(3), torch.randn(3))

# Load the saved model back and call it in-place.
with torch.compiler._fullgraph_package(mode="load", path="/tmp/model.pt2"):
    f(torch.randn(3), torch.randn(3))
```
Internally, fullgraph_package will call export and aoti step by step and finally package all the compiled models into a single package when we exit the context. (see test_fullgraph_package.py for more information)

Since this is an AOT workflow, we have to assume that the package file(s) saved from this API are likely to be used in a different environment, despite the fact that it may or may not cause soundness issue. To make this BC-safe, we need to give each compilation a unique name so that the models being loaded are addressable from a different process.

(Right now we just assign the unique compilatin name/id using the function's FQN. Ideally user should be able to specify the compilation name directly on torch.compile())

For now, the plan of record is we want to make this ctx manager API standalone because we need to put at least three arguments here:
1. `mode` which indicates whether we're saving or loading the packages
2. `path` to specify where we store the compiled package.
3. `frontend` to speciy whether the programming model.

We haven't made a final decision on where to put this funtionality, and the purpose of this diff is to demonstrate what an ideal workflow to do partial model capture for torchnative.

Test Plan:
buck test mode/opt caffe2/test:test_export -- -r FullgraphPackage

OSS: test_fullgraph_package.py

Differential Revision: D67353701




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

zhxchen17 commented:
@ezyang Does this look something in the right direction?

ezyang commented:
The API code sample looks in the right direction. Checking with Horace if he wants to bikeshed any. I haven't looked at impl yet.

zhxchen17 committed:
Eager style export V0 API. (#143432)

Summary:

Prototype of an end-to-end export workflow to call a torch.compiled model eagerly and package every single compiled model in the wrapped region of the code.

Code sample:

```
compile(fullgraph=True)
def f(x, y):
    return x + y

# Compile the model and save it on disk
with torch.compiler.fullgraph_package(mode="package", path="/tmp/model.pt2", frontend="dynamo"):
    f(torch.randn(3), torch.randn(3))

# Load the saved model back and call it in-place.
with torch.compiler.fullgraph_package(mode="load", path="/tmp/model.pt2"):
    f(torch.randn(3), torch.randn(3))
```
Internally, fullgraph_package will call export and aoti step by step and finally package all the compiled models into a single package when we exit the context. (see test_fullgraph_package.py for more information)

Since this is an AOT workflow, we have to assume that the package file(s) saved from this API are likely to be used in a different environment, despite the fact that it may or may not cause soundness issue. To make this BC-safe, we need to give each compilation a unique name so that the models being loaded are addressable from a different process.

(Right now we just assign the unique compilatin name/id using the function's FQN. Ideally user should be able to specify the compilation name directly on torch.compile())

For now, the plan of record is we want to make this ctx manager API standalone because we need to put at least three arguments here:
1. `mode` which indicates whether we're saving or loading the packages
2. `path` to specify where we store the compiled package.
3. `frontend` to speciy whether the programming model.

We haven't made a final decision on where to put this funtionality, and the purpose of this diff is to demonstrate what an ideal workflow to do partial model capture for torchnative.

Test Plan:
buck test mode/opt caffe2/test:test_export -- -r FullgraphPackage

OSS: test_fullgraph_package.py

Differential Revision: D67353701

Chillee commented:
Main questions/thoughts:

1. How does this handle weight sharing? For example, in prefill/decode I want the same model's weights: https://github.com/pytorch-labs/gpt-fast/blob/main/generate.py#L68 
2. How does this handle modules?

zhxchen17 commented:
x-link nn.Module input support: https://fburl.com/2fwxcrko

zhxchen17 commented:
(Moving to D68459341)
